### PR TITLE
[NVPTX] Add support for f16 fabs

### PR DESCRIPTION
**Pull request by AlexMaclean**
Add support for f16 and f16x2 support for abs. See PTX ISA 9.7.4.6. Half Precision Floating Point Instructions: abs https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#half-precision-floating-point-instructions-abs

### CODE REVIEW

File: llvm/test/CodeGen/NVPTX/f16-abs.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/f16-abs.ll:83]: Huh. This is interesting. If we are allowed to do `fabs` via sign masking, perhaps instead of promotion to `fp32` for scalars, we should cutom-lower it to a logical op for the fallback path for fp16 scalars where native fp16 abs is not supported. That's likely more efficient than abs.f32 plus two conversions.

**Comment by justinfargnoli** [comment]
Feel free to ignore [comment](https://github.com/llvm/llvm-project/pull/116107#discussion_r1841203171) and [comment](https://github.com/llvm/llvm-project/pull/116107#discussion_r1841210789). I'll file issues for them.

**Comment by justinfargnoli** [comment]
Overall, this LGTM. I'm holding off on approving this in case @Artem-B wants you to address [comment](https://github.com/llvm/llvm-project/pull/116107#discussion_r1841210178) before merging.

File: llvm/test/CodeGen/NVPTX/f16-abs.ll
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/f16-abs.ll:83]: To clarify -- it's an optimization opportunity. If it easy to incorporate into this change -- great. If not, can be done separately. This CL is good to go as is.

Comment (AlexMaclean) [reply llvm/test/CodeGen/NVPTX/f16-abs.ll:83]: I think it is fine from the perspective of llvm IR's semantics to implement f16 abs with an and is fine, it's probably more conformant because it will preserve NaN payloads, while the conversions may not. That being said, I'm not sure about the perf implications of going this route. Maybe in some cases maybe the abs could be strung together with other promoted operations and result in better codegen?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:868]: I don't even think we support this architecture.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:877]: I don't think this is necessary, in fact I suspect this whole case can be removed. `hasBF16Math` already checks for sm_80 and since sm_80 is not supported until ptx_70 we can assume this requirement will be met as well.

https://github.com/llvm/llvm-project/blob/ec066d30e29fce388b1722971970d73ec65f14fb/llvm/lib/Target/NVPTX/NVPTXSubtarget.h#L86

File: llvm/test/CodeGen/NVPTX/f16-abs.ll
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/f16-abs.ll:83]: I think the only case where it will matter is when we would have a cluster of fp16 ops promoted to f32, with a few `abs` in the middle. If LLVM would decide to do that abs in fp16, that indeed may be slower. I'm not sure how exectly this is handled. In theory it would boil down to cost analysis between `two casts + and` vs `fp32 fabs`, with `fp32 fabs` winning.

That may be something to teach instcombine about, if we do not, yet.

In either case, sm_60 and older GPUs are nearly obsolete these days and are not worth spending much effort on. We can just leave things as is.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Done, I've removed the for loop

File: llvm/test/CodeGen/NVPTX/f16-abs.ll
Comment (AlexMaclean) [reply llvm/test/CodeGen/NVPTX/f16-abs.ll:83]: Sounds good, will leave as is for older architectures for now.

**Review by justinfargnoli — APPROVED** [review]

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (justinfargnoli) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: @Artem-B is it correct that we don't support architectures older than sm_20?

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Yes. sm_20 is the oldest supported variant. We should officially deprecate it and start removing support for it, as NVIDIA dropped support for it in 2018 (https://nvidia.custhelp.com/app/answers/detail/a_id/4654).

File: llvm/test/CodeGen/NVPTX/f16-abs.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/f16-abs.ll:3]: Few nits on the tests -- I would explicitly set PTX version high enough on all test cases to make sure that native `abs` instruction is disabled by the GPU type or by explicit attributes, and not unintentionally so by the too-low default PTX version.

File: llvm/test/CodeGen/NVPTX/f16-abs.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/f16-abs.ll:35]: ptx 6.5 enables `abs` for `fp16`, but not `bf16`. That may be another useful test case to include.

**Review by Artem-B — APPROVED** [review]
LGTM in general + few test nits.

File: llvm/test/CodeGen/NVPTX/f16-abs.ll
Comment (AlexMaclean) [reply llvm/test/CodeGen/NVPTX/f16-abs.ll:35]: This test only covers f16 abs. bf16 abs was already supported and has coverage in the bf16 tests. Should I add bf16 cases to this test as well?

File: llvm/test/CodeGen/NVPTX/f16-abs.ll
Comment (AlexMaclean) [reply llvm/test/CodeGen/NVPTX/f16-abs.ll:3]: Okay, I've updated the cases to include each of the following:
- no f16 abs due to insufficient hardware
- no f16 abs due to insufficient ptx version
- no f16 abs due to explicit disable
- f16 abs supported

Comment (AlexMaclean) [reply llvm/test/CodeGen/NVPTX/f16-abs.ll]: Apologies, I did not, I've just pushed 0c7af950b7dbefcb3ee545ddd74670ffcb6d7869

**Review by Artem-B — APPROVED** [review]

**Pull request merged**
